linux/devdoc/completion_port_linux_requirements.md

```diff
@@ -0,0 +1,187 @@
+```
+
+If `completion_port` is `NULL`, `completion_port_remove` shall return.
+
+If `socket` is `INVALID_SOCKET`, `completion_port_remove` shall return.
+
+`completion_port_remove` shall remove the underlying socket from the epoll by calling `epoll_ctl` with `EPOLL_CTL_DEL`.
+
+### epoll_worker_func
+
+```c
+static int epoll_worker_func(void* parameter)
+```
+
+If `parameter` is `NULL`, `epoll_worker_func` shall do nothing.
+
+`epoll_worker_func` shall call `epoll_wait` to wait for an epoll event to become signaled with a timeout of 2 Seconds.
+
+If the events value contains `EPOLLRDHUP` (hang up), `epoll_worker_func` shall the following:
+
+- `epoll_worker_func` shall receive the `EPOLL_THREAD_DATA` value from the ptr variable from the `epoll_event` data ptr.
+
+- Then call the `event_callback` callback with the `event_callback_ctx` and `COMPLETION_PORT_RESULT_ABANDONED`.
+
+- Then remove the `EPOLL_THREAD_DATA` from the list and free the object
+
+If the events value contains `EPOLLIN` (recv), `epoll_worker_func` shall do the following:
+
+- `epoll_worker_func` shall receive the `EPOLL_THREAD_DATA` value from the ptr variable from the `epoll_event` data ptr.
+
+- Then `epoll_worker_func` shall call `recv` and do the following:
+
+- If the recv size < 0, then:
+
+  - If `errno` is either `EAGAIN` or `EWOULDBLOCK`, then `epoll_worker_func` shall break and call `event_callback` callback with the `event_callback_context` with `COMPLETION_PORT_RESULT_OK`
+
+  - If `errno` is `ECONNRESET`, then `epoll_worker_func` shall call the `event_callback` callback with the `event_callback_context` and `COMPLETION_PORT_RESULT_ABANDONED`
+
+- If the recv size equal 0, then `epoll_worker_func` shall call `event_callback` callback with the `event_callback_context` and `COMPLETION_PORT_RESULT_OK`
+
+- If the recv size > 0, then if recv size < buffer length is less than `epoll_worker_func` shall issue another recv, otherwise we shall call `event_callback` callback with the `event_callback_context` with `COMPLETION_PORT_RESULT_OK`
+
+- Then remove the `EPOLL_THREAD_DATA` from the list and free the object
+
+If the events value contains `EPOLLOUT` (send), `epoll_worker_func` shall the following:
+
+- `epoll_worker_func` shall receive the `EPOLL_THREAD_DATA` value from the ptr variable from the `epoll_event` data ptr.
+
+- `epoll_worker_func` shall loop through the total buffers and send the data.
+
+- if send returns value is < 0 and not `EAGAIN` or `EWOULDBLOCK`, then `epoll_worker_func` shall do the following:
+
+  - if `errno` is `ECONNRESET`, then `on_send_complete` shall be called with `COMPLETION_PORT_RESULT_ABANDONED`.
+
+  - if `errno` is anything else, then `on_send_complete` shall be called with `COMPLETION_PORT_RESULT_ERROR`.
+
+- Then remove the `EPOLL_THREAD_DATA` from the list and free the object
+
+If the thread_access_cnt variable is not 0, `epoll_worker_func` shall issue another `epoll_wait`, otherwise it shall exit.
```

Comment:
>thread_access_cnt [](http://example.com/codeflow?start=7&length=17)

This only appears here. It is impossible to guess what this will be and it actually looks like this is part of the internal design of the module.
This should be fully removed IMO.